src/platform/riot/coap_udp_ip_channel.c

```diff
@@ -223,89 +254,68 @@ static lf_ret_t CoapUdpIpChannel_try_connect(NetworkChannel *untyped_self) {
-  LF_DEBUG(NET, "CoapUdpIpChannel: Client close connection callback");
+  (void)remote; // This pointer is useless and always NULL
+  COAP_UDP_IP_CHANNEL_DEBUG("Client close connection callback");
   (void)memo;
   (void)pdu;
-  (void)remote;
 
   // Do nothing
 }
 
 static void CoapUdpIpChannel_close_connection(NetworkChannel *untyped_self) {
-  LF_DEBUG(NET, "CoapUdpIpChannel: Close connection");
+  COAP_UDP_IP_CHANNEL_DEBUG("Close connection");
   CoapUdpIpChannel *self = (CoapUdpIpChannel *)untyped_self;
 
   // Immediately close the channel
-  mutex_lock(&self->state_mutex);
-  { _CoapUdpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_CLOSED); }
-  mutex_unlock(&self->state_mutex);
+  _CoapUdpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_CLOSED);
 
   // Inform the other federate that the channel is closed
   _CoapUdpIpChannel_send_coap_message(&self->remote, "/disconnect", _CoapUdpIpChannel_client_close_connection_callback);
 }
 
-static void _CoapUdpIpChannel_client_send_blocking_callback(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
-                                                            const sock_udp_ep_t *remote) {
-  LF_DEBUG(NET, "CoapUdpIpChannel: Client send blocking callback");
-  (void)memo;
-  (void)pdu;
-  (void)remote;
+static void _client_send_blocking_callback(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                                           const sock_udp_ep_t *remote) {
+  (void)remote; // This pointer is useless and always NULL
+  COAP_UDP_IP_CHANNEL_DEBUG("Client send blocking callback");
+  CoapUdpIpChannel *self = _CoapUdpIpChannel_get_coap_channel_by_remote(&memo->remote_ep);
 
-  // Do nothing
+  if (memo->state == GCOAP_MEMO_TIMEOUT) {
+    // Failure
+    COAP_UDP_IP_CHANNEL_ERR("TIMEOUT");
+    _CoapUdpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_LOST_CONNECTION);
+  } else if (coap_get_code_class(pdu) != COAP_CLASS_SUCCESS) {
+    // Failure
+    COAP_UDP_IP_CHANNEL_ERR("CONNECTION REJECTED");
+    _CoapUdpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_LOST_CONNECTION);
+  }
+
+  self->send_ack_received = true;
 }
 
 static lf_ret_t CoapUdpIpChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
-  LF_DEBUG(NET, "CoapUdpIpChannel: Send blocking");
+  COAP_UDP_IP_CHANNEL_DEBUG("Send blocking");
   CoapUdpIpChannel *self = (CoapUdpIpChannel *)untyped_self;
 
   // Send message
-  if (_CoapUdpIpChannel_send_coap_message_with_payload(self, &self->remote, "/message",
-                                                       _CoapUdpIpChannel_client_send_blocking_callback, message)) {
-    return LF_OK;
+  self->send_ack_received = false;
+  if (_CoapUdpIpChannel_send_coap_message_with_payload(self, &self->remote, "/message", _client_send_blocking_callback,
+                                                       message)) {
+    // Wait until the response handler confirms the ack or times out
+    while (!self->send_ack_received) {
```

Comment:
I guess this is the only way for us to guarantee in-order reliable delivery, right?

Reply:
Actually this was implemented to achieve a `blocking` send. Just like in TCP we want to send blocking so we wait for the message to be `acked`.

But you are correct this somewhat also gives us in-order guarantee, but I am not sure if it covers all cases. We can think about this more later I think.

Reply:
Yes, this is a sane starting point and we can think about edge cases and intermittent connections etc later